### PR TITLE
feat(url): refactor module logic movig some logic to UrlHandler Component

### PR DESCRIPTION
**Pull request by tajespasarela**
EX-4864

The purpose of this PR is to refactor the URL module trying to solve future problems with multiple Load events and handling the URL state params and extra params.

Basically I moved most of the logic to the `UrlHandler` component, having this the responsibility of communication with browser URL.

Load and popstate events are emitted from the Component with the parameters already parsed.

Then there are two new emitters: `PushableUrlStateChanged` and `ReplaceableUrlStateChanged`. The component responds to each one replacing or pushing a state to browser URL.

I remove the config from the state since it is no longer needed. Only needed in the component.
I remove unnecessary actions and getters.

So the 2 direction flows are the following:

Load URL -> UrlHandler emits `ParamsLoadedFromUrl` -> the rest of the modules listen this (included URL module)

URL states changes -> Emitters emit `PushableUrlStateChanged` or `ReplaceableUrlStateChanged` -> UrlHandler modifies the browser URL (if the final URL is the same as the current this change is not made).

### CODE REVIEW

File: packages/x-components/src/views/Layout.vue [hunk 1]
```diff
@@ -396,6 +396,7 @@
   import BaseSuggestions from '../components/suggestions/base-suggestions.vue';
   import { infiniteScroll } from '../directives/infinite-scroll/infinite-scroll';
   import { XInstaller } from '../x-installer/x-installer';
+  // eslint-disable-next-line max-len
```
Comment (pmareke) [inline packages/x-components/src/views/Layout.vue:399]: You can remove this eslint rule.

File: packages/x-components/src/x-modules/related-tags/store/actions/set-url-params-from-the.url.action.ts [hunk 2]
```diff
@@ -34,12 +33,12 @@ const createRelatedTags = (relatedTags: string[], query: string): RelatedTag[] =
  * @public
  */
 export const setUrlParamsFromTheUrl: RelatedTagsXStoreModule['actions']['setUrlParamsFromTheUrl'] =
-  ({ commit }, { query, relatedTag }: Dictionary<UrlParamValue>) => {
-    const newRelatedTag = relatedTag as string[];
-    const newQuery = query as string;
+  ({ commit }, { query, relatedTag }: UrlParams) => {
+    const newRelatedTags = relatedTag;
```
Comment (pmareke) [inline packages/x-components/src/x-modules/related-tags/store/actions/set-url-params-from-the.url.action.ts:37]: I think it's `newRelatedTag` as you received the new one.

File: packages/x-components/src/x-modules/related-tags/store/actions/set-url-params-from-the.url.action.ts [hunk 3]
```diff
@@ -34,12 +33,12 @@ const createRelatedTags = (relatedTags: string[], query: string): RelatedTag[] =
...
+    const newQuery = query;
 
-    if (newRelatedTag) {
-      commit('setSelectedRelatedTags', createRelatedTags(newRelatedTag, newQuery));
+    if (newRelatedTags) {
```
Comment (pmareke) [inline packages/x-components/src/x-modules/related-tags/store/actions/set-url-params-from-the.url.action.ts:40]: Same here.

File: packages/x-components/src/x-modules/related-tags/store/actions/set-url-params-from-the.url.action.ts (hunk 2 above)
Comment (pmareke) [reply packages/x-components/src/x-modules/related-tags/store/actions/set-url-params-from-the.url.action.ts:37]: Also this const it's not longer needed, previously it was necessary in order to cast from unknown to string.

File: packages/x-components/src/x-modules/related-tags/store/actions/set-url-params-from-the.url.action.ts [hunk 4]
```diff
@@ -34,12 +33,12 @@ const createRelatedTags = (relatedTags: string[], query: string): RelatedTag[] =
  * @public
  */
 export const setUrlParamsFromTheUrl: RelatedTagsXStoreModule['actions']['setUrlParamsFromTheUrl'] =
-  ({ commit }, { query, relatedTag }: Dictionary<UrlParamValue>) => {
-    const newRelatedTag = relatedTag as string[];
-    const newQuery = query as string;
+  ({ commit }, { query, relatedTag }: UrlParams) => {
+    const newRelatedTags = relatedTag;
+    const newQuery = query;
```
Comment (pmareke) [inline packages/x-components/src/x-modules/related-tags/store/actions/set-url-params-from-the.url.action.ts:38]: Same with the query.

File: packages/x-components/src/x-modules/url/components/__tests__/url-handler.spec.ts [hunk 5]
```diff
@@ -16,23 +16,27 @@ import { UrlHandler } from '../index';
...
-    },
-    {
-      localVue
-    }
-  );
+  const wrapperTemplate = mount({ template, components: { UrlHandler } }, { localVue });
   const wrapper = wrapperTemplate.findComponent(UrlHandler);
-  const $x = wrapperTemplate.vm.$x;
+
+  function getUrlFunction(event: string) {
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/components/__tests__/url-handler.spec.ts:22]: If you're dispatching an event I'l rename the method, right now confused more than anything else.

File: packages/x-components/src/x-modules/url/components/url-handler.vue [hunk 6]
```diff
@@ -1,82 +1,209 @@
...
   })
   export default class UrlHandler extends Vue {
-    // eslint-disable-next-line @typescript-eslint/no-empty-function
-    render(): void {}
+    /**
+     * Computed to know which params we must to get from URL. It gets the params names from the
+     * initial state, to get all default params names, and also from the `$attrs` to get the extra
+     * params names to take into account.
+     *
+     * @returns An array with the name of the params.
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/components/url-handler.vue:20]: Add blank line below.

File: packages/x-components/src/x-modules/url/components/url-handler.vue [hunk 7]
```diff
@@ -1,82 +1,209 @@
...
+    }
 
     /**
-     * Saves the new parameter names, if present, and add two XEvents to the
-     * (`onload` and `onpopstate`) listeners of the browser window.
+     * Returns the mapping of the param keys used in the URL is configured through $attrs. This way
+     * we can support any param and extra param, no matters its name.
+     *
+     * @param paramName - The param name to get the Url key.
+     * @returns A dictionary where the key is the internal param name (example: 'query') and the
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/components/url-handler.vue:32]: It returns a string not a Dictionary, please review the doc of the method.

File: packages/x-components/src/x-modules/url/components/url-handler.vue [hunk 8]
```diff
@@ -1,82 +1,209 @@
...
+      });
     }
 
     /**
-     * Open X if there is a query in the Url.
+     * Returns the URL param value parsed depending on its type in the initial store state. As we
+     * can not know what type can have an extra param, all extra params are parsed as strings. We
+     * know if it is an extra param because it is not in the initial state.
      *
+     * @param urlSearchParams - The {@link URLSearchParams} to get the param.
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/components/url-handler.vue:176]: Add new lines please.

File: packages/x-components/src/x-modules/url/events.types.ts [hunk 9]
```diff
@@ -8,33 +7,18 @@ import { UrlGetters } from './store/types';
...
   /**
-   * The configuration for handling the URL has changed.
-   * * Payload: The new URL config.
+   * Url loaded from the url changed.
+   * * Payload: The new URL params.
    */
-  UrlConfigProvided: UrlConfig;
+  ParamsLoadedFromUrl: UrlParams;
   /**
    * Url state changed.
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/events.types.ts:15]: Review this comment is the same as the `ReplaceableUrlStateChanged`.

File: packages/x-components/src/x-modules/url/store/__tests__/actions.spec.ts [hunk 10]
```diff
@@ -1,137 +1,26 @@
...
-
-      await store.dispatch('updateUrl');
-
-      // eslint-disable-next-line @typescript-eslint/unbound-method
-      expect(window.history.replaceState).toHaveBeenCalled();
-    });
-  });
 
   describe('updateStoreFromUrl', () => {
     it('should update the state with the correct url parameters', async () => {
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/store/__tests__/actions.spec.ts:9]: Why did you remove 3 tests (push, replace an invalid params in the url)? I'd rollback and leave them.

File: packages/x-components/src/x-modules/url/store/emitters.ts [hunk 11]
```diff
@@ -1,15 +1,29 @@
...
  *
  * @internal
  */
 export const urlEmitters = createStoreEmitters(urlXStoreModule, {
-  UrlStateChanged: (_, getters) => getters.urlParams,
-  UrlChanged: {
+  PushableUrlStateChanged: {
+    selector: (_, getters) => getters.urlParams,
+    filter: (newValues: UrlParams, oldValues: UrlParams) =>
+      !pushableParams.some(key => oldValues[key] && oldValues[key] !== newValues[key])
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/store/emitters.ts:22]: Can we extract this logic into a smaller method?

File: packages/x-components/src/x-modules/url/store/module.ts [hunk 12]
```diff
@@ -1,38 +1,38 @@
...
+ * The initial state of the Url store module. This is exported and used in other parts of the code
+ * to use as default values for {@link UrlState}.
+ *
+ * @internal
+ */
+export const initialUrlState: UrlState = {
+  params: {
+    query: '',
+    page: 1,
+    filter: [],
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/store/module.ts:15]: We're using a singular to represent a list of things.

File: packages/x-components/src/x-modules/url/store/module.ts [hunk 13]
```diff
@@ -1,38 +1,38 @@
...
+    sort: '',
+    scroll: 0,
+    relatedTag: []
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/store/module.ts:18]: Same here.

File: packages/x-components/src/x-modules/url/store/module.ts [hunk 14]
```diff
@@ -42,21 +42,17 @@ export const urlXStoreModule: UrlXStoreModule = {
     setQuery(state, query) {
       state.params.query = query;
     },
-    setRelatedTags(state, relatedTag) {
-      state.params.relatedTag = relatedTag;
+    setRelatedTags(state, relatedTags) {
+      state.params.relatedTag = relatedTags.map(relatedTag => relatedTag.tag);
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/store/module.ts:46]: This `relatedTag = relatedTags` is so weird.

File: packages/x-components/src/x-modules/url/store/module.ts [hunk 15]
```diff
@@ -42,21 +42,17 @@ export const urlXStoreModule: UrlXStoreModule = {
...
     },
     setFilters(state, newFilters) {
-      state.params.filters = newFilters;
+      state.params.filter = newFilters.map(filter => filter.id as string);
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/store/module.ts:49]: Same here.

**Review by pmareke — CHANGES_REQUESTED** [review]

File: packages/x-components/src/views/Layout.vue (hunk 1 above)
Comment (tajespasarela) [reply packages/x-components/src/views/Layout.vue:397]: done

File: packages/x-components/src/x-modules/url/store/__tests__/actions.spec.ts (hunk 10 above)
Comment (tajespasarela) [reply packages/x-components/src/x-modules/url/store/__tests__/actions.spec.ts:10]: The remove emtpy values is responsibility of the getter, I added a test for it there

File: packages/x-components/src/x-modules/url/store/__tests__/getters.spec.ts [hunk 16]
```diff
@@ -26,23 +20,51 @@ describe('testing url module getters', () => {
...
+    });
+
+    expect(store.getters.urlParams).toEqual<Partial<UrlGetters['urlParams']>>({
+      query: 'doramion',
+      warehouse: '11111'
+    });
+
+    resetUrlStateWith(store, { params: { warehouse: '' } });
+
+    expect(store.getters.urlParams).toEqual<Partial<UrlGetters['urlParams']>>({
```
Comment (tiborux) [inline packages/x-components/src/x-modules/url/store/__tests__/getters.spec.ts:66]: The test is failling because the getter is returning the warehouse: "". In the getter, the warehouse is inside the params, so you are not cleaning the extra param when they are empty: 

image

File: packages/x-components/src/x-modules/url/components/url-handler.vue [hunk 17]
```diff
@@ -121,14 +245,14 @@ In this example, the `UrlHandler` component changes the following query paramete
...
-- `filters` to be `f`
+- `filter` to be `f`
 - `sort` to be `s`
 
 _Try to make some requests and take a look to the url!_
 
 ```vue
 <template>
-  <UrlHandler :query="query" :page="page" :filters="filters" :sort="sort" />
+  <UrlHandler query="query" page="page" filter="filter" sort="sort" />
```
[the remark this answers is not in the document]
Comment (tajespasarela) [reply packages/x-components/src/x-modules/url/components/url-handler.vue:258]: fixed

File: packages/x-components/src/x-modules/url/components/url-handler.vue [hunk 18]
```diff
@@ -1,82 +1,209 @@
...
+      paramName: string
+    ): UrlParamValue | undefined {
+      const key = this.getUrlKey(paramName);
+      if (!(paramName in initialUrlState.params)) {
+        return urlSearchParams.get(key) ?? undefined;
+      } else {
+        switch (typeof initialUrlState.params[paramName]) {
+          case 'number': {
+            const numberOrNull = urlSearchParams.get(key);
+            return numberOrNull ? Number(numberOrNull) : undefined;
```
[the remark this answers is not in the document]
Comment (tajespasarela) [reply packages/x-components/src/x-modules/url/components/url-handler.vue]: we have a rule to avoid the use of it:
image

File: packages/x-components/src/x-modules/url/components/url-handler.vue [hunk 19]
```diff
@@ -1,82 +1,209 @@
...
+     * Gets the {@link UrlParams} from the URL, including only the params defined by
+     * {@link paramsNames}.
+     *
+     * @returns UrlParams obtained from URL.
+     * @internal
+     */
+    protected getUrlParams(): UrlParams {
+      const urlSearchParams = new URL(window.location.href).searchParams;
+      const urlParams = this.paramsNames.reduce((urlParams, paramName) => {
+        urlParams[paramName] = this.getParamByType(urlSearchParams, paramName);
```
[the remark this answers is not in the document]
Comment (tajespasarela) [reply packages/x-components/src/x-modules/url/components/url-handler.vue]: I buy you the `if` with the `has` but the other changes...

You are simplifying a method to complicate the other. I prefer to leave in one single method the responsibility of getting one URL param

[the remark this answers is not in the document]
Comment (tajespasarela) [reply packages/x-components/src/x-modules/url/components/url-handler.vue]: also is not a good practice to cast types with `!` if not necessary. TypeScript is here to help, if you silence him, then it makes no sense.

File: packages/x-components/src/x-modules/history-queries/store/actions/set-url-params.action.ts [hunk 20]
```diff
@@ -0,0 +1,19 @@
+import { HistoryQueriesXStoreModule } from '../types';
+
+/**
+ * Default implementation for the {@link QuerySuggestionsXStoreModule.setUrlParams}.
```
Comment (tiborux) [inline packages/x-components/src/x-modules/history-queries/store/actions/set-url-params.action.ts:4]: HistoryQueriesXStoreModule.setUrlParams

File: packages/x-components/src/x-modules/identifier-results/store/actions/set-url-params.action.ts [hunk 21]
```diff
@@ -0,0 +1,19 @@
+import { IdentifierResultsXStoreModule } from '../types';
+
+/**
+ * Default implementation for the {@link QuerySuggestionsXStoreModule.setUrlParams}.
```
Comment (tiborux) [inline packages/x-components/src/x-modules/identifier-results/store/actions/set-url-params.action.ts:4]: IdentifierResultsXStoreModule.setUrlParams

File: packages/x-components/src/x-modules/url/components/url-handler.vue (hunk 19 above)
[the remark this answers is not in the document]
Comment (guillei10) [reply packages/x-components/src/x-modules/url/components/url-handler.vue]: I know and you are right but in this particular case, once you know `urlSearchParams.has(param)` is `true` then you are 100% certain `urlSearchParams.get(param)` will not be null. But Typescript doesn't know that.

Actually now I realise the return of the boolean `case` is also wrong. The `?` is opening the possibility to return `undefined`, but the function is not typed to return undefined. I would use `!` here also because again, you are 100% sure `urlSearchParams.get(param)` will return a value (a string actually)

[the remark this answers is not in the document]
Comment (guillei10) [reply packages/x-components/src/x-modules/url/components/url-handler.vue]: Another doubt I have. Is there a preference to use?

`return Object.assign({}, initialUrlState.params, urlParams);`
 instead of 
`return {...initialUrlState.params, ...urlParams}`

File: packages/x-components/src/x-modules/url/store/__tests__/getters.spec.ts (hunk 16 above)
Comment (tiborux) [reply packages/x-components/src/x-modules/url/store/__tests__/getters.spec.ts:67]: ` it('removes all the empty extra params', () => {
    resetUrlStateWith(store, {
      params: {
        query: 'doramion'
      },
      extraParams: {
        warehouse: '11111'
      }
    });

    expect(store.getters.urlParams).toEqual<Partial<UrlGetters['urlParams']>>({
      query: 'doramion',
      warehouse: '11111'
    });

    resetUrlStateWith(store, {
      params: {
        query: 'doramion'
      },
      extraParams: { warehouse: '' }
    });

    expect(store.getters.urlParams).toEqual<Partial<UrlGetters['urlParams']>>({
      query: 'doramion'
    });`

File: packages/x-components/src/x-modules/url/store/__tests__/getters.spec.ts [hunk 22]
```diff
@@ -26,23 +20,51 @@ describe('testing url module getters', () => {
...
+      query: 'doramion'
     });
   });
```
Comment (tiborux) [inline packages/x-components/src/x-modules/url/store/__tests__/getters.spec.ts]:
```suggestion
  it('removes all the empty extra params', () => {
    resetUrlStateWith(store, {
      params: {
        query: 'doramion'
      },
      extraParams: {
        warehouse: '11111'
      }
    });

    expect(store.getters.urlParams).toEqual<Partial<UrlGetters['urlParams']>>({
      query: 'doramion',
      warehouse: '11111'
    });

    resetUrlStateWith(store, {
      params: {
        query: 'doramion'
      },
      extraParams: { warehouse: '' }
    });

    expect(store.getters.urlParams).toEqual<Partial<UrlGetters['urlParams']>>({
      query: 'doramion'
    });
  });
```

**Review by pmareke — CHANGES_REQUESTED** [review]
Fantastic job, sorry for the facets service mistake. Apart from @tiborux's comments I'll just rename the helper method in the url spec file.

File: packages/x-components/src/x-modules/url/components/__tests__/url-handler.spec.ts (hunk 5 above)
Comment (tajespasarela) [reply packages/x-components/src/x-modules/url/components/__tests__/url-handler.spec.ts:22]: done

File: packages/x-components/src/x-modules/url/components/__tests__/url-handler.spec.ts [hunk 23]
```diff
@@ -19,7 +19,7 @@ function renderUrlHandler({ template = `<UrlHandler />` }: UrlHandlerOptions = {
   const wrapperTemplate = mount({ template, components: { UrlHandler } }, { localVue });
   const wrapper = wrapperTemplate.findComponent(UrlHandler);
 
-  function getUrlFunction(event: string) {
+  function dispatchWindowEvent(event: string) {
```
Comment (pmareke) [inline packages/x-components/src/x-modules/url/components/__tests__/url-handler.spec.ts:22]: Please add the return type.

**Review by pmareke — CHANGES_REQUESTED** [review]

File: packages/x-components/temp/x-components.api.md [hunk 24]
```diff
@@ -0,0 +1,4421 @@
+## API Report File for "@empathyco/x-components"
```
Comment (pmareke) [inline packages/x-components/temp/x-components.api.md:1]: Remove this file please.

File: packages/x-components/src/x-modules/history-queries/wiring.ts [hunk 25]
```diff
@@ -46,6 +46,13 @@ export const addQueryToHistoryQueries = wireDispatch('addQueryToHistory');
  */
 export const setHistoryQueriesQuery = wireCommit('setQuery');
 
+/**
+ * Sets the next queries state `query` from url.
```
Comment (pmareke) [inline packages/x-components/src/x-modules/history-queries/wiring.ts:50]: We're in the history queries module, please update the comment.

**Review by pmareke — APPROVED** [review]
